### PR TITLE
Enable Boundedness Checks for Spectrahedra and Intersections with Unbounded Components

### LINKED ISSUE #16663: Check boundedness of `Intersection` of convex sets
As discussed with @RussTedrake and @mpetersen94, currently the boundedness check for the `Intersection` of convex sets throws unless any of the intersected sets are bounded. But it can be that even if all of the sets are unbounded the intersection is bounded. Here I report the math (might need some cleaning) that I believe we should implement for this check.

- The problem of determining if the intersection of convex sets is bounded is equal to the problem of checking if a convex set of the form `S = {x: g_i(x) <= 0 for all i}` is bounded.
- We form the recession `K` cone of `S`. This will have the form `K = {x: h_i(x) <= 0 for all i}`. To construct the functions `h_i` from the `g_i` we will need a separate method for each `ConvexSet`. For example, if a set is a polyhedron, then `g_i(x) = a' x - b` and `h_i(x) = a' x`. For an `Ellipsoid` the cone `K` is the nullspace of the Hessian. Etc.
- The set `S` is bounded iff `K` contains a nonzero point.
- First we check if `K` is pointed. If not, then `S` is unbounded. One way to do this is to check if the linearizations of the `h_i(x)` at `x = 0` form a basis for `R^n`.
- If `K` is pointed then we have `h_i(x) = 0` for all `i` iff `x = 0`.
- Finally, we then check if `K` contains a point `x != 0` by solving the convex program: `find x, y` such that ` h_i(x) <= y for all i`, `y <= 0`, and `sum_k y_k <= - 1`.
- If `K` has a nonzero point `x` in it, then the problem above is feasible. This because `x != 0` implies  `h_i(x) < 0` for some `i`.
- If `K = {0}` then the above problem is infeasible.

Note: This is essentially the dual of the check we currently do to see if a polyhedron is bounded. For a polyhedron `{x: A x <= b}` the check of pointedness of `K` becomes `ker(A) = {0}`. And the convex program becomes the LP `find x, y` such that ` A x <= y`, `y <= 0`, and `sum_k y_k <= - 1`. When we implement the stuff above, I'd also modify the check for the polyhedron since I think this primal LP is much easier to interpret than the dual check we are currently doing.

### PR DESCRIPTION
**Pull request by cohnt**
This implements `DoIsBoundedShortcut()` for `ConvexSet`, and also modifies the implementations for `Intersection` and `Spectahedron` to use the generic base class method (after checking some edge cases).

Fixes #16663 (although potentially not in the most efficient way). +@russtedrake for feature review.

The implementation is slightly different than what I suggested in [this comment](https://github.com/RobotLocomotion/drake/issues/16663#issuecomment-1619017030) -- instead of maximizing and minimizing each dimension, I maximize the displacement between two feasible points along each dimension. This doubles the number of decision variables, but halves the number of optimization problems solved.

- - -
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19888)

### CODE REVIEW

**Comment by RussTedrake** [comment]
Let's ask +@TobiaMarcucci for a concept review, too... since he opened the issue and thought about it.

**Comment by TobiaMarcucci** [comment]
I think that if at some point we'll move to a unified conic description of the the sets in the GCS this check should be easy (i.e. a single cone problem with size almost equal to the size of the set). Currently it's hard to find an elegant and simple way to check boundedness of any convex set. So I guess the question is when is a good time to change basic representation of the convex sets.

I think this approach is simple and useful for debugging, although as Tommy says not very efficient.

The idea of maximizing the displacement makes sense if setting up the problems is expensive. Otherwise solving two problems with half the variables is cheaper than solving one problem with double the variables. For example, say that solving a convex problem takes n^2 time, the first approach takes 2n^2, while the second approach takes (2n)^2=4n^2. Since you are only updating the cost coefficients between one problem and the next, I'm not sure which approach works best.

**Comment by cohnt** [comment]
I like Tobia's idea of moving towards a unified conic description. But I expect that will take some time to land, and being able to check the boundedness of an intersection is already relevant to Drake users. For example, we'll need it in order to add a "debug mode" to GCS (which I'm sure @shaoyuancc would like to have).

This fix "works" (pending feature review, platform review, tests, etc.) now, and we can always replace it later with a more efficient implementation.

Re maximizing and minimizing each dimension, or maximizing the displacement, I don't have a preference between the two, and I can quickly switch the code around to do one or the other. Note that this code hopefully wouldn't be called very often. If all of the convex sets in an intersection are bounded, the result is returned as bounded without going into this logic.

**Comment by TobiaMarcucci** [comment]
I agree. And especially if the interface won't change when we move to the conic representation, there is no damage in merging now a less performant way to do the check.

File: geometry/optimization/spectrahedron.cc [hunk 1]
```diff
@@ -77,8 +77,18 @@ std::unique_ptr<ConvexSet> Spectrahedron::DoClone() const {
 }
 
 bool Spectrahedron::DoIsBounded() const {
-  throw std::runtime_error(
-      "Spectrahedron::IsBounded() is not implemented yet.");
+  // CSDP cannot handle an unconstrained spectrahedron (i.e. a set whose only
```
Comment (TobiaMarcucci) [inline geometry/optimization/spectrahedron.cc:80]: I'm not familiar with CSDP, but this seems strange? I would imagine that the important distinction is between bounded and unbounded, rather than constrained and unconstrained. For example, does CSDP accept something like `X >= 0` and `X[0,0] == 1`? (This would be constrained but unbounded.)

**Review by TobiaMarcucci — COMMENTED** [review]
I've some observations, but overall it looks good!

File: geometry/optimization/spectrahedron.cc (hunk 1 above)
Comment (cohnt) [reply geometry/optimization/spectrahedron.cc:80]: There's an error in either CSDP or the Drake interface to CSDP. It will give there error:
```
Problem must have at least one constraint.
unknown file: Failure
C++ exception with description "CsdpSolver: the CSDP library exited via a fatal exception" thrown in the test body.
```

**Review by cohnt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19888)* status: 4 unresolved discussions, LGTM missing from assignees TobiaMarcucci,RussTedrake(platform)

___
*[`geometry/optimization/convex_set.cc` line 49 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/19888#gh-1281551277:-NarP0k35X2R2pM7QHGo:b-896fix) ([raw file](https://github.com/RobotLocomotion/drake/blob/36a147b931ca6f2feb779941ff24e0557afd7463/geometry/optimization/convex_set.cc#L49)):*
<details><summary><i>Previously, TobiaMarcucci (Tobia Marcucci) wrote…</i></summary><blockquote>

If there is no empirical performance reason, I would say that it is better to solve 2n problems with n variables rather than n problems with 2n variables.
</blockquote></details>

Done.
___
*[`geometry/optimization/spectrahedron.cc` line 85 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/19888#gh-1281561709:-NarP20LEc9CLCkIlPSZ:bcit5j3) ([raw file](https://github.com/RobotLocomotion/drake/blob/36a147b931ca6f2feb779941ff24e0557afd7463/geometry/optimization/spectrahedron.cc#L85)):*
<details><summary><i>Previously, TobiaMarcucci (Tobia Marcucci) wrote…</i></summary><blockquote>

This check seems potentially brittle? What if by mistake I added twice the constraint that `X >= 0`? Would this check fail?
</blockquote></details>

This check is purely to handle the problem where CSDP refuses to consider problems without any constraints (besides the builtin PSD constraints). As long as the optimization problem can be run, it will be caught in `ConvexSet::DoIsBounded`. (And I've added a test case to specifically examine this.)
___
*[`geometry/optimization/test/spectrahedron_test.cc` line 325 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/19888#gh-1281563086:-NarPNTx-s-MUljYsqjF:bcyaomw) ([raw file](https://github.com/RobotLocomotion/drake/blob/36a147b931ca6f2feb779941ff24e0557afd7463/geometry/optimization/test/spectrahedron_test.cc#L325)):*
<details><summary><i>Previously, TobiaMarcucci (Tobia Marcucci) wrote…</i></summary><blockquote>

By the way, do we think that some tests of `ConvexSet::DoIsBounded` belong to the `convex_set_test.cc` file?
</blockquote></details>

Just added a bunch of test cases to make it a little more robust. I'm not as familiar with the corner cases of SDPs, but if you have specific suggestions, I'm happy to add them.

It's hard to put `ConvexSet::DoIsBounded` tests in `convex_set_test.cc`, since all subclasses have their own implementations, and you can't construct a pure `ConvexSet`. `Intersection` and `Spectrahedron` are the only ones that ever call the base class version, so I think we have to test it through them.

**Review by RussTedrake — COMMENTED** [review]
:lgtm: pending @TobiaMarcucci 's review.  Thanks!

> Reviewed 4 of 6 files at r1, 1 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-:-Nasy8vk0Dec0G0WfKhI:bydhifa)* status: 5 unresolved discussions, LGTM missing from assignee TobiaMarcucci

___
*[`geometry/optimization/convex_set.cc` line 61 at r3](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-NasxC3n9-UEnTij9-PT:-NasxC3n9-UEnTij9-PU:b-fkxbx6) ([raw file](https://github.com/RobotLocomotion/drake/blob/c96dc4c87398ad86adad9015dcb63f59ba6bc4db/geometry/optimization/convex_set.cc#L61)):*
> ```cpp
>   VectorXd objective_vector(ambient_dimension());
>   for (int i = 0; i < ambient_dimension(); ++i) {
>     objective_vector = VectorXd::Zero(ambient_dimension());
> ```

btw -- you could just call `objective_vector.setZero()` here.
___
*[`geometry/optimization/spectrahedron.cc` line 80 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/19888#gh-1281559423:-Nasxdea7X0m-84tbp1e:bg054yr) ([raw file](https://github.com/RobotLocomotion/drake/blob/36a147b931ca6f2feb779941ff24e0557afd7463/geometry/optimization/spectrahedron.cc#L80)):*
<details><summary><i>Previously, cohnt (Thomas Cohn) wrote…</i></summary><blockquote>

There's an error in either CSDP or the Drake interface to CSDP. It will give there error:
```
Problem must have at least one constraint.
unknown file: Failure
C++ exception with description "CsdpSolver: the CSDP library exited via a fatal exception" thrown in the test body.
```
</blockquote></details>

could we please open an issue for that with a minimal test case?

**Review by cohnt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19888)* status: 4 unresolved discussions, LGTM missing from assignee TobiaMarcucci

___
*[`geometry/optimization/convex_set.cc` line 61 at r3](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-NasxC3n9-UEnTij9-PT:-NavxgWP2hsNo5czPjiz:b-896fix) ([raw file](https://github.com/RobotLocomotion/drake/blob/c96dc4c87398ad86adad9015dcb63f59ba6bc4db/geometry/optimization/convex_set.cc#L61)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

btw -- you could just call `objective_vector.setZero()` here.
</blockquote></details>

Done.
___
*[`geometry/optimization/spectrahedron.cc` line 80 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/19888#gh-1281559423:-Navxf0V9f6-XU7_tdun:b-cc0k4) ([raw file](https://github.com/RobotLocomotion/drake/blob/36a147b931ca6f2feb779941ff24e0557afd7463/geometry/optimization/spectrahedron.cc#L80)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

could we please open an issue for that with a minimal test case?
</blockquote></details>

Done. #19927

**Review by TobiaMarcucci — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-:-NaxtlYu4V0QehteTHNj:bnfp4nl)* status: :shipit: complete! all discussions resolved, LGTM from assignees TobiaMarcucci,RussTedrake(platform)

**Review by RussTedrake — COMMENTED** [review]
+@sherm1 for platform review, please.

> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-:-Nb9TIYX8buS3tOAGzhY:bppyoqq)* status: LGTM missing from assignee sherm1(platform)

**Review by sherm1 — COMMENTED** [review]
Platform :lgtm: with one suggestion to consider

> Reviewed 4 of 6 files at r1, 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-:-NbFkZYW1S2-o26tThWp:bwi5r1m)* status: 2 unresolved discussions

___
*[`geometry/optimization/convex_set.cc` line 42 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-NbFfBPwDn3k2V-JzaAB:-NbFfBPx3D_ADcWW7A94:bu61gb5) ([raw file](https://github.com/RobotLocomotion/drake/blob/cea99e45cd347755b7abfa2d1b0f1f1cde52518d/geometry/optimization/convex_set.cc#L42)):*
> ```cpp
> }
> 
> bool ConvexSet::DoIsBounded() const {
> ```

BTW it's awkward to have a base class virtual implementation that has to be invoked by derived classes for correct behavior. If I've understood the intent correctly, here's an alternate design sketch for your consideration:
```c++
bool ConvexSet::IsBounded() const {
  if (std::optional<bool> result = DoIsBoundedShortcut())
    return *result;
  return IsBoundedFullCheck();
}
virtual std::optional<bool> ConvexSet::DoIsBoundedShortcut() 
  { return std::nullopt; }
```
This has the advantage that derived classes only have to think about themselves -- do I have a quick test for boundedness? If not the base class takes responsibility for the test.
___
*[`geometry/optimization/convex_set.cc` line 50 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-NbFchK3BrJSw-dVGb-M:-NbFchK40VlkJjdit2e9:bnz6g0c) ([raw file](https://github.com/RobotLocomotion/drake/blob/cea99e45cd347755b7abfa2d1b0f1f1cde52518d/geometry/optimization/convex_set.cc#L50)):*
> ```cpp
>     return true;
>   }
>   // Iteratively try to maximize maximize the distance between x and y along
> ```

typo: maximize maximize -> maximize

**Review by cohnt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19888)* status: 1 unresolved discussion

___
*[`geometry/optimization/convex_set.cc` line 42 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-NbFfBPwDn3k2V-JzaAB:-NbFnUrqC3cFf366bzvy:b6uvnri) ([raw file](https://github.com/RobotLocomotion/drake/blob/cea99e45cd347755b7abfa2d1b0f1f1cde52518d/geometry/optimization/convex_set.cc#L42)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW it's awkward to have a base class virtual implementation that has to be invoked by derived classes for correct behavior. If I've understood the intent correctly, here's an alternate design sketch for your consideration:
```c++
bool ConvexSet::IsBounded() const {
  if (std::optional<bool> result = DoIsBoundedShortcut())
    return *result;
  return IsBoundedFullCheck();
}
virtual std::optional<bool> ConvexSet::DoIsBoundedShortcut() 
  { return std::nullopt; }
```
This has the advantage that derived classes only have to think about themselves -- do I have a quick test for boundedness? If not the base class takes responsibility for the test.

</blockquote></details>

The way it's currently set up is that `ConvexSet::DoIsBounded` is valid for any derived class of `ConvexSet`, but most classes override it with their own, more efficient checks. The only classes that ever have to call the expensive check implemented here are `Intersection` and `Spectrahedron`, but even then we can sometimes avoid it. All tests still pass if you comment out the subclass-specific behavior in `Intersection` and `Spectrahedron` and just call `ConvexSet::DoIsBounded` right away. (Note that `Spectrahedron` wouldn't pass without Mosek, due to the issue with CSDP described in #19927. This edge case is handled in `Spectrahedron`'s implementation of `DoIsBounded`.)
___
*[`geometry/optimization/convex_set.cc` line 50 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-NbFchK3BrJSw-dVGb-M:-NbFnTiAEFrVFyr_usaL:b-896fix) ([raw file](https://github.com/RobotLocomotion/drake/blob/cea99e45cd347755b7abfa2d1b0f1f1cde52518d/geometry/optimization/convex_set.cc#L50)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

typo: maximize maximize -> maximize
</blockquote></details>

Done.

**Review by sherm1 — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19888)* status: 1 unresolved discussion

___
*[`geometry/optimization/convex_set.cc` line 42 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-NbFfBPwDn3k2V-JzaAB:-NbFsTqf52zRZDXvl8nH:bce5tdk) ([raw file](https://github.com/RobotLocomotion/drake/blob/cea99e45cd347755b7abfa2d1b0f1f1cde52518d/geometry/optimization/convex_set.cc#L42)):*
<details><summary><i>Previously, cohnt (Thomas Cohn) wrote…</i></summary><blockquote>

The way it's currently set up is that `ConvexSet::DoIsBounded` is valid for any derived class of `ConvexSet`, but most classes override it with their own, more efficient checks. The only classes that ever have to call the expensive check implemented here are `Intersection` and `Spectrahedron`, but even then we can sometimes avoid it. All tests still pass if you comment out the subclass-specific behavior in `Intersection` and `Spectrahedron` and just call `ConvexSet::DoIsBounded` right away. (Note that `Spectrahedron` wouldn't pass without Mosek, due to the issue with CSDP described in #19927. This edge case is handled in `Spectrahedron`'s implementation of `DoIsBounded`.)
</blockquote></details>

That would still be true if you replace DoIsBounded() with DoIsBoundedShortcut() but the code would better reflect the intent, IMO. As written it took me a while to figure out what was going on (derived classes are given a chance to shortcut the IsBounded check if they know how, otherwise a default expensive method is used). You could also clarify that with better comments on the DoIsBounded() virtual, directed at implementers of derived classes. But to me that doesn't seem as good as making the code clearly match the intent, and leaving responsibility for the "no shortcut available" case to the base class.

**Review by cohnt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19888)* status: :shipit: complete! all discussions resolved, LGTM from assignees TobiaMarcucci,sherm1(platform),RussTedrake(platform)

___
*[`geometry/optimization/convex_set.cc` line 42 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-NbFfBPwDn3k2V-JzaAB:-NbG1QhT6lpV2OZ10QoH:b-yyu82i) ([raw file](https://github.com/RobotLocomotion/drake/blob/cea99e45cd347755b7abfa2d1b0f1f1cde52518d/geometry/optimization/convex_set.cc#L42)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

That would still be true if you replace DoIsBounded() with DoIsBoundedShortcut() but the code would better reflect the intent, IMO. As written it took me a while to figure out what was going on (derived classes are given a chance to shortcut the IsBounded check if they know how, otherwise a default expensive method is used). You could also clarify that with better comments on the DoIsBounded() virtual, directed at implementers of derived classes. But to me that doesn't seem as good as making the code clearly match the intent, and leaving responsibility for the "no shortcut available" case to the base class. 
</blockquote></details>

Done.

I also added some further comments on `ConvexSet::IsBounded` to clarify this behavior, and pointed them to `Intersection` and `Spectrahedron` as examples of how to use this.

Will this require another round of feature review?

**Review by sherm1 — COMMENTED** [review]
> Reviewed 1 of 1 files at r5, 5 of 5 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19888)* status: 3 unresolved discussions

___
*[`geometry/optimization/convex_set.h` line 89 at r6](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-NbGFjwy0uVNyGvdY8ld:-NbGFjwy0uVNyGvdY8le:b-igkynf) ([raw file](https://github.com/RobotLocomotion/drake/blob/53a6c5e23f8f3df1ee0b5355d02228165fe313dc/geometry/optimization/convex_set.h#L89)):*
minor: this is a great comment, but out of place here in the public Doxygen since it is targeted to an implementer rather than an end user. Should move to the protected DoIsBoundedShortcut() since that's what the derived class writers will have to implement.

I am confused now by this comment:
> Derived classes may also use the generic base-class implementation of
  DoIsBounded. 

The derived classes should either provide a shortcut or not, right? If not then the _base class_ is responsible for the generic boundedness check, not the derived class.
___
*[`geometry/optimization/convex_set.h` line 263 at r6](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-NbGGLf1CIYShbTjdzDR:-NbGGLf1CIYShbTjdzDS:b-djx195) ([raw file](https://github.com/RobotLocomotion/drake/blob/53a6c5e23f8f3df1ee0b5355d02228165fe313dc/geometry/optimization/convex_set.h#L263)):*
> ```objc
>   /** Non-virtual interface implementation for IsBounded().
>   @pre ambient_dimension() >= 0 */
>   virtual bool DoIsBounded() const;
> ```

I'm confused now why this is still a virtual method. It should only be invoked by the base class so should just be a private function.

I think this should just be renamed to something like `GenericIsBoundedCheck()` and made private. Then any derived classes that previously overrode DoIsBounded() should override DoIsBoundedShortcut() instead.

I was hoping to simplify the implementation not make it more complicated -- it's not an improvement to have two ways to do the same thing.

BTW it will be easy to find the places that need changing since they will fail to compile as soon as DoIsBounded() gets a new name.
___
*[`geometry/optimization/convex_set.cc` line 42 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-NbFfBPwDn3k2V-JzaAB:-NbGHPuB2SIkgCmO57w2:b-rn5e2f) ([raw file](https://github.com/RobotLocomotion/drake/blob/cea99e45cd347755b7abfa2d1b0f1f1cde52518d/geometry/optimization/convex_set.cc#L42)):*
> Will this require another round of feature review?

Should be a straightforward translation that won't need more feature review. But that assumes virtual DoIsBoundedShortcut() _replaces_ the DoIsBounded() virtual. If they both exist as virtuals then the logic is too mysterious for me!
___
*[`geometry/optimization/convex_set.cc` line 50 at r6](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-NbGI2F1ByjXzba-yeAW:-NbGI2F22hycqem8l7Jg:b-y7xdr5) ([raw file](https://github.com/RobotLocomotion/drake/blob/53a6c5e23f8f3df1ee0b5355d02228165fe313dc/geometry/optimization/convex_set.cc#L50)):*
> ```cpp
>     return true;
>   }
>   // Iteratively try to minimize or maximize the an x in the convex set for
> ```

minor: this sentence doesn't scan any more -- some words missing?

**Review by sherm1 — COMMENTED** [review]
:lgtm_cancel: temporarily for some logic untangling

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-:-NbGK4xM9bRbyHSgs3c3:b3p57l0)* status: 3 unresolved discussions, LGTM missing from assignee sherm1(platform)

**Review by cohnt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19888)* status: 3 unresolved discussions, LGTM missing from assignee sherm1(platform)

___
*[`geometry/optimization/convex_set.h` line 89 at r6](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-NbGFjwy0uVNyGvdY8ld:-NbGWJ2h3sZI-Ai2FLI4:b-hlhvct) ([raw file](https://github.com/RobotLocomotion/drake/blob/53a6c5e23f8f3df1ee0b5355d02228165fe313dc/geometry/optimization/convex_set.h#L89)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

minor: this is a great comment, but out of place here in the public Doxygen since it is targeted to an implementer rather than an end user. Should move to the protected DoIsBoundedShortcut() since that's what the derived class writers will have to implement.

I am confused now by this comment:
> Derived classes may also use the generic base-class implementation of
  DoIsBounded. 

The derived classes should either provide a shortcut or not, right? If not then the _base class_ is responsible for the generic boundedness check, not the derived class.
</blockquote></details>

Moved to the protected `DoIsBoundedShortcut`, and removed the note about `DoIsBounded`, since I've fully replaced it with `DoIsBoundedShortcut`.
___
*[`geometry/optimization/convex_set.h` line 263 at r6](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-NbGGLf1CIYShbTjdzDR:-NbGWSWE2Y9Arv5FUV9X:b-77qsl0) ([raw file](https://github.com/RobotLocomotion/drake/blob/53a6c5e23f8f3df1ee0b5355d02228165fe313dc/geometry/optimization/convex_set.h#L263)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

I'm confused now why this is still a virtual method. It should only be invoked by the base class so should just be a private function.

I think this should just be renamed to something like `GenericIsBoundedCheck()` and made private. Then any derived classes that previously overrode DoIsBounded() should override DoIsBoundedShortcut() instead.

I was hoping to simplify the implementation not make it more complicated -- it's not an improvement to have two ways to do the same thing.

BTW it will be easy to find the places that need changing since they will fail to compile as soon as DoIsBounded() gets a new name.
</blockquote></details>

I've created a unified `DoIsBoundedShortcut` that all subclasses of `ConvexSet` use, and renamed this to `GenericIsBoundedCheck` and moved it to private.

Derived classes which had their own boundedness checks now use `DoIsBoundedShortcut`, and simply always take the shortcut.
___
*[`geometry/optimization/convex_set.cc` line 50 at r6](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-NbGI2F1ByjXzba-yeAW:-NbGWwNAC-2t5rX8rt_v:b-8dwh3h) ([raw file](https://github.com/RobotLocomotion/drake/blob/53a6c5e23f8f3df1ee0b5355d02228165fe313dc/geometry/optimization/convex_set.cc#L50)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

minor: this sentence doesn't scan any more -- some words missing?
</blockquote></details>

Should be clearer now.

**Review by sherm1 — COMMENTED** [review]
Beautiful! Thanks for your patience, Thomas. Platform :lgtm:.

> Reviewed 15 of 16 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19888#-:-NbGiX9m8S7Y9cmASLan:bwup0qi)* status: :shipit: complete! all discussions resolved, LGTM from assignees TobiaMarcucci,sherm1(platform),RussTedrake(platform)

**Comment by cohnt** [comment]
@sherm1 this failed the CI, are you sure it was okay to merge it? The issue was somewhere in `//bindings/pydrake/visualization:py/model_visualizer_test`, which my code shouldn't have touched at all. (It also builds correctly on my machine, but I'm not on Mac.) See: https://drake-jenkins.csail.mit.edu/job/mac-arm-ventura-clang-bazel-experimental-release/1559/consoleFull

**Comment by sherm1** [comment]
I saw that but interpreted it as a Mac flake since it was so remote from this work (and Mac CI is not required for merging). It might cause trouble later but I think not.

**Comment by cohnt** [comment]
Got it! Just checking to make sure
